scipy/linalg/_solvers.py

```diff
@@ -416,23 +441,66 @@ def solve_discrete_are(a, b, q, r):
             raise ValueError("Matrix {} should be symmetric/hermitian."
                              "".format("qr"[ind]))
 
+    # Check if the generalized case is required with omitted arguments
+    # perform late shape checking etc. 
+    generalized_case = e is not None or s is not None
+    e_is_I_or_None = True
+    if generalized_case:
+        if e is not None:
+            e = np.atleast_2d(_asarray_validated(e,check_finite=True))
+            if not np.equal(*e.shape):
+                raise ValueError("Matrix e should be square.")
+            if m != e.shape[0]:
+                raise ValueError("Matrix a and e should have the same shape.")
+            # numpy.linalg.cond doesn't check for exact zeros and
+            # emits a runtime warning. Hence the following manual check.
+            min_sv = svd(e, compute_uv=False)[-1]
+            if min_sv == 0. or min_sv < np.spacing(1.)*norm(e,1):
+                raise ValueError('Matrix e is numerically singular')
+            e_is_I_or_None = False
+        if s is not None:
+            s = np.atleast_2d(_asarray_validated(s, check_finite=True))
+            if s.shape != b.shape:
+                raise ValueError("Matrix b and s should have the same shape.")
+
     # Form the matrix pencil        
     H = np.zeros((2*m + n, 2*m + n), dtype = r_or_c)
     H[:m, :m] = a
     H[:m, 2*m:] = b
     H[m:2*m, :m] = -q
-    H[m:2*m, m:2*m] = np.eye(m)
+    H[m:2*m, m:2*m] = np.eye(m) if e is None else e.T
+    H[m:2*m, 2*m:] = 0. if s is None else -s
+    H[2*m:, :m] = 0. if s is None else s.T
     H[2*m:, 2*m:] = r
 
     J = np.zeros_like(H, dtype = r_or_c)
-    J[:m, :m] = np.eye(m)
+    J[:m, :m] = np.eye(m) if e is None else e
     J[m:2*m, m:2*m] = a.conj().T
     J[2*m:, m:2*m] = -b.conj().T
 
+    if balanced:
+        # xGEBAL does not remove the diagonals before scaling. Also 
+        # to avoid destroying the Symplectic structure, we follow Ref.3
+        M = np.abs(H) + np.abs(J)
+        M[np.diag_indices_from(M)] = 0.
+        _, (sca, _) = balance(M,separate=1,permute=0)
+        # do we need to bother? 
+        if not np.allclose(sca,np.ones_like(sca)):
```

Comment:
maybe use explicit ``rtol, atol`` here. since ``allclose`` and ``assert_allclose`` have different default, I can never remember the values.